### PR TITLE
Replace recursive with loop in PackedValuesBlockHash

### PR DESCRIPTION
**Pull request by dnhatn**
This change replaces the recursion with a loop when packing and hashing multiple keys. While the recursive version is clever, it may not be as straightforward for future readers. Using a loop also helps us avoid StackOverflow when grouping by a large number of keys.

### CODE REVIEW

**Comment by dnhatn** [comment]
@nik9000 Please hold off the review. This doesn't pass all our tests.

**Comment by dnhatn** [comment]
@nik9000 I think this is ready for review. Would you please take a look? Thank you!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java [hunk 1]
```diff
@@ -79,19 +76,19 @@ class AddWork extends LongLongBlockHash.AbstractAddBlock {
         final BatchEncoder[] encoders = new BatchEncoder[groups.size()];
         final int[] positionOffsets = new int[groups.size()];
         final int[] valueOffsets = new int[groups.size()];
-        final BytesRef[] scratches = new BytesRef[groups.size()];
+        final BytesRef scratch = new BytesRef();
+        final int[] loopedIndices = new int[groups.size()];
+        final int[] valueCounts = new int[groups.size()];
+        final int[] bytesStarts = new int[groups.size()];
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java:82]: I think we'd be better of making an object for each group and have an array of those. But whatever works for you.

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java:82]: ++, will do.

Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java]: I pushed https://github.com/elastic/elasticsearch/pull/99992/commits/f41c6e83f1c03f92c021c1d1a8d3ded3bbd0d659, which is much cleaner.

**Comment by dnhatn** [comment]
@nik9000 Thank you for the review.

**Comment by dnhatn** [comment]
@elasticsearchmachine update branch

**Comment by dnhatn** [comment]
update branch

**Comment by dnhatn** [comment]
@elasticmachine test this please

**Pull request merged**